lib/concurrent/struct/abstract_struct.rb

```diff
@@ -0,0 +1,141 @@
+module Concurrent
+
+  # @!visibility private
+  module AbstractStruct
+
+    # @!macro [attach] struct_length
+    #
+    #   Returns the number of struct members.
+    #
+    #   @return [Fixnum] the number of struct members
+    def length
+      self.class::MEMBERS.length
+    end
+    alias_method :size, :length
+
+    # @!macro [attach] struct_members
+    #
+    #   Returns the struct members as an array of symbols.
+    #
+    #   @return [Array] the struct members as an array of symbols
+    def members
+      self.class::MEMBERS.dup
+    end
+
+    protected
+
+    # @!macro struct_values
+    #
+    # @!visibility private
+    def ns_values
+      @values.dup
+    end
+
+    # @!macro struct_values_at
+    #
+    # @!visibility private
+    def ns_values_at(indexes)
+      @values.values_at(*indexes)
+    end
+
+    # @!macro struct_to_h
+    #
+    # @!visibility private
+    def ns_to_h
+      length.times.reduce({}){|memo, i| memo[self.class::MEMBERS[i]] = @values[i]; memo}
+    end
+
+    # @!macro struct_get
+    #
+    # @!visibility private
+    def ns_get(member)
+      if member.is_a? Integer
+        if member >= @values.length
+          raise IndexError.new("offset #{member} too large for struct(size:#{@values.length})")
+        end
+        @values[member]
+      else
+        send(member)
+      end
+    rescue NoMethodError
+      raise NameError.new("no member '#{member}' in struct")
+    end
+
+    # @!macro struct_equality
+    #
+    # @!visibility private
+    def ns_equality(other)
+      self.class == other.class && self.values == other.values
+    end
+
+    # @!macro struct_each
+    #
+    # @!visibility private
+    def ns_each
+      values.each{|value| yield value }
```

Comment:
The subclasses do that since some of the subclasses are synchronized and others aren't. I assumed (perhaps incorrectly) that the `#to_enum` call should occur outside of `#synchronize`.